pkg/controller/networkpolicy/networkpolicy_controller.go

```diff
@@ -1296,8 +1324,38 @@ func (n *NetworkPolicyController) syncAddressGroup(key string) error {
 		internalNP := internalNPObj.(*antreatypes.NetworkPolicy)
 		addrGroupNodeNames = addrGroupNodeNames.Union(internalNP.SpanMeta.NodeNames)
 	}
+	n.addressGroupMutex.Lock()
```

Comment:
This will serialize most work of addressGroupWorkers, the existing benchmark tests can show its impact.
Could you explain the race condition here? Let's see if it's possible to avoid locking it.

Reply:
Im thinking of the following scenario:
1. A CG update occurs
2. A CNP referencing this CG is created/updated

The first event causes the syncInternalGroup() to be triggered causing a re-processing of affected CNPs wherein the CG is updated with the new selector info in createAddressGroupForClusterGroupCRD().
Concurrently the syncAddressGroup() could be triggered because of CNP event, to update the Span of the AddressGroup.

I am trying to think how to avoid this

Reply:
done by removing the need to set selector in AddressGroups created for internal Groups as the group member computation will be done by internal group.. so no need to set the Selector for such cases